skimage/morphology/rolling_ball.py

```diff
@@ -113,68 +121,43 @@ def rolling_ball(image, radius=50, white_background=False):
-    img_original = img.copy()
-    img_size = img.shape
-    x = img
-    stride = (1, 1)
-
-    # pad the image
-    padding_amount = kernel_size // 2 * 2
-    half_pad = padding_amount // 2
-    img = np.Inf * np.ones(x.shape + padding_amount, dtype=x.dtype)
-    img[half_pad[0]:-half_pad[0], half_pad[1]:-half_pad[1]] = x
-    large_img_size = img.shape
-
-    # the following three blocks implement a variant of 2d convolution
-    # in python. Each time a kernel is applied to a window, sagitta is added
-    # to the respective pixel values and the reduction is done using `min` instead
-    # of `sum`
-
-    # window of affected pixel indices relative to anchor
-    x_idx = np.arange(kernel_size[1])
-    y_idx = large_img_size[1] * np.arange(kernel_size[0])
-    kernel_idx = (x_idx[np.newaxis, :] + y_idx[:, np.newaxis]).flatten()
-
-    # indices of each ancor for a kernel
-    # (top left corner instead of center)
-    x_anchors = np.arange(
-        large_img_size[1] - kernel_size[1] + 1, step=stride[1])
-    y_anchors = large_img_size[1] * \
-        np.arange(large_img_size[0] - kernel_size[0] + 1, step=stride[0])
-    anchor_offsets = (x_anchors[np.newaxis, :] +
-                      y_anchors[:, np.newaxis]).flatten()
-
-    # compute px indices in the image and apply the function/kernel
-    # large images or kernel sizes don't fit into memory
-    # do it in batches instead
-    background = np.zeros(img_original.size)
-    batch_size = int(2 ** 10)
+    img = np.pad(img, spacial_upper_bound,
+                 constant_values=np.iinfo(img.dtype).max)
+
+    background = np.zeros_like(image)
     flat_img = img.flatten()
-    flat_sagitta = sagitta.flatten()
-    flat_kernel = kernel.flatten()
-    for low in range(0, len(anchor_offsets), batch_size):
-        high = np.minimum(low + batch_size, len(anchor_offsets))
-        filter_idx = anchor_offsets[low:high,
-                                    np.newaxis] + kernel_idx[np.newaxis, :]
+    flat_sagitta = sagitta.flatten()[np.newaxis, :]
+    flat_kernel = kernel.flatten()[np.newaxis, :]
+
+    windowed = view_as_windows(img, kernel.shape)
+    chunk_anchors = [range(0, shape, step)
+                     for shape, step in zip(image.shape, chunk_size)]
+    y_step, x_step = chunk_size
+
+    for y, x in product(*chunk_anchors):
+        chunk = windowed[y:y+y_step, x:x+x_step]
+        chunk_shape = chunk.shape[:2]
+        chunk = chunk.reshape((-1, kernel.size))
         background_partial = np.min(
-            (flat_img[filter_idx] + flat_sagitta[np.newaxis, :]) * flat_kernel[np.newaxis, :], axis=1)
-        background[low:high] = background_partial
-
-    background = background.reshape(img_size).astype(img_original.dtype)
-    filtered_image = img_original - background
+            (chunk + flat_sagitta) * flat_kernel, axis=-1)
+        background_partial = background_partial.reshape(chunk_shape)
+        background[y:y+y_step, x:x+x_step] = background_partial
```

Comment:
So many red lines in this commit — I love it! 😃

Reply:
thank you. Even more red lines in the last commit. I decided to refactor it into two functions, which makes input validation much cleaner. Also, the critical part of the code is now in cython, which makes it very memory efficient; I'm happy for any ideas that make it faster :)